File: src/main/java/com/erdidev/timemanager/service/TaskService.java
package com.erdidev.timemanager.service;

import com.erdidev.timemanager.dto.TaskDto;
import com.erdidev.timemanager.exception.CategoryNotFoundException;
import com.erdidev.timemanager.exception.ProjectNotFoundException;
import com.erdidev.timemanager.exception.TaskNotFoundException;
import com.erdidev.timemanager.mapper.TaskMapper;
import com.erdidev.timemanager.model.*;
import com.erdidev.timemanager.repository.*;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

@Slf4j
@Service
@RequiredArgsConstructor
public class TaskService {
    private final TaskRepository taskRepository;
    private final CategoryRepository categoryRepository;
    private final ProjectRepository projectRepository;
    private final TaskAttachmentRepository attachmentRepository;
    private final TaskMapper taskMapper;

    @Transactional(readOnly = true)
    public Page<TaskDto> getTasks(Pageable pageable) {
        log.debug("Fetching tasks page: {}", pageable);
        return taskRepository.findAll(pageable)
                .map(taskMapper::toDto);
    }

    @Transactional(readOnly = true)
    public Page<TaskDto> getTasksByProject(Long projectId, Pageable pageable) {
        log.debug("Fetching tasks for project: {}", projectId);
        return taskRepository.findByProjectId(projectId, pageable)
                .map(taskMapper::toDto);
    }

    @Transactional(readOnly = true)
    public Page<TaskDto> getTasksByCategory(Long categoryId, Pageable pageable) {
        log.debug("Fetching tasks for category: {}", categoryId);
        return taskRepository.findByCategoryId(categoryId, pageable)
                .map(taskMapper::toDto);
    }

    @Transactional(readOnly = true)
    public List<TaskDto> getTasksByStatus(TaskStatus status) {
        log.debug("Fetching tasks with status: {}", status);
        return taskRepository.findByStatus(status).stream()
                .map(taskMapper::toDto)
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public List<TaskDto> getTasksByPriority(Priority priority) {
        log.debug("Fetching tasks with priority: {}", priority);
        return taskRepository.findByPriority(priority).stream()
                .map(taskMapper::toDto)
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public List<TaskDto> getOverdueTasks() {
        log.debug("Fetching overdue tasks");
        return taskRepository.findByDueDateBefore(LocalDateTime.now()).stream()
                .map(taskMapper::toDto)
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public TaskDto getTask(Long id) {
        log.debug("Fetching task with id: {}", id);
        return taskRepository.findById(id)
                .map(taskMapper::toDto)
                .orElseThrow(() -> new TaskNotFoundException(id));
    }

    @Transactional
    public TaskDto createTask(TaskDto taskDto) {
        log.debug("Creating task: {}", taskDto);
        Task task = taskMapper.toEntity(taskDto);
        
        // Set default status and priority if not provided
        task.setStatus(taskDto.getStatus() != null ? taskDto.getStatus() : TaskStatus.TODO);
        task.setPriority(taskDto.getPriority() != null ? taskDto.getPriority() : Priority.MEDIUM);
        
        // Assign to category if provided
        if (taskDto.getCategoryId() != null) {
            Category category = categoryRepository.findById(taskDto.getCategoryId())
                    .orElseThrow(() -> new CategoryNotFoundException(taskDto.getCategoryId()));
            task.setCategory(category);
        }
        
        // Assign to project if provided
        if (taskDto.getProjectId() != null) {
            Project project = projectRepository.findById(taskDto.getProjectId())
                    .orElseThrow(() -> new ProjectNotFoundException(taskDto.getProjectId()));
            task.setProject(project);
        }
        
        Task savedTask = taskRepository.save(task);
        return taskMapper.toDto(savedTask);
    }

    @Transactional
    public TaskDto updateTask(Long id, TaskDto taskDto) {
        log.debug("Updating task with id: {}", id);
        Task existingTask = taskRepository.findById(id)
                .orElseThrow(() -> new TaskNotFoundException(id));
        
        taskMapper.updateEntity(taskDto, existingTask);
        
        // Update category if provided
        if (taskDto.getCategoryId() != null) {
            Category category = categoryRepository.findById(taskDto.getCategoryId())
                    .orElseThrow(() -> new CategoryNotFoundException(taskDto.getCategoryId()));
            existingTask.setCategory(category);
        }
        
        // Update project if provided
        if (taskDto.getProjectId() != null) {
            Project project = projectRepository.findById(taskDto.getProjectId())
                    .orElseThrow(() -> new ProjectNotFoundException(taskDto.getProjectId()));
            existingTask.setProject(project);
        }
        
        Task updatedTask = taskRepository.save(existingTask);
        return taskMapper.toDto(updatedTask);
    }

    @Transactional
    public void deleteTask(Long id) {
        log.debug("Deleting task with id: {}", id);
        if (!taskRepository.existsById(id)) {
            throw new TaskNotFoundException(id);
        }
        // Delete associated attachments first
        attachmentRepository.deleteByTaskId(id);
        taskRepository.deleteById(id);
    }

    @Transactional(readOnly = true)
    public List<TaskDto> searchTasks(String query) {
        log.debug("Searching tasks with query: {}", query);
        return taskRepository.findByTitleContainingIgnoreCase(query).stream()
                .map(taskMapper::toDto)
                .collect(Collectors.toList());
    }
} 

File: frontend/time-manager-ui/src/services/taskService.ts
import axios from 'axios';

const API_URL = 'http://localhost:8081/api/v1';

export interface TaskDto {
  id?: number;
  title: string;
  description?: string;
  status?: TaskStatus;
  priority?: Priority;
  dueDate?: string;
  estimatedHours?: number;
  actualHours?: number;
  projectId?: number;
  categoryId?: number;
  creatorId?: number;
  ownerId?: number;
  assigneeId?: number;
  createdAt?: string;
  updatedAt?: string;
}

export enum TaskStatus {
  TODO = 'TODO',
  IN_PROGRESS = 'IN_PROGRESS',
  DONE = 'DONE',
  BLOCKED = 'BLOCKED',
  REVIEW = 'REVIEW'
}

export enum Priority {
  LOW = 'LOW',
  MEDIUM = 'MEDIUM',
  HIGH = 'HIGH',
  URGENT = 'URGENT'
}

export interface PageResponse<T> {
  content: T[];
  pageable: {
    pageNumber: number;
    pageSize: number;
    sort: {
      sorted: boolean;
      unsorted: boolean;
      empty: boolean;
    }
  };
  totalElements: number;
  totalPages: number;
  last: boolean;
  first: boolean;
  size: number;
  number: number;
  sort: {
    sorted: boolean;
    unsorted: boolean;
    empty: boolean;
  };
  numberOfElements: number;
  empty: boolean;
}

const apiClient = axios.create({
  baseURL: API_URL,
  withCredentials: true,
});

const taskService = {
  getTasks: async (page = 0, size = 10, sort = 'createdAt,desc'): Promise<PageResponse<TaskDto>> => {
    const response = await apiClient.get(`/tasks?page=${page}&size=${size}&sort=${sort}`);
    return response.data;
  },

  getTask: async (id: number): Promise<TaskDto> => {
    const response = await apiClient.get(`/tasks/${id}`);
    return response.data;
  },

  createTask: async (task: TaskDto): Promise<TaskDto> => {
    const response = await apiClient.post('/tasks', task);
    return response.data;
  },

  updateTask: async (id: number, task: TaskDto): Promise<TaskDto> => {
    const response = await apiClient.put(`/tasks/${id}`, task);
    return response.data;
  },

  deleteTask: async (id: number): Promise<void> => {
    await apiClient.delete(`/tasks/${id}`);
  },

  searchTasks: async (query: string): Promise<TaskDto[]> => {
    const response = await apiClient.get(`/tasks/search?query=${encodeURIComponent(query)}`);
    return response.data;
  },

  getTasksByProject: async (projectId: number, page = 0, size = 10, sort = 'createdAt,desc'): Promise<PageResponse<TaskDto>> => {
    const response = await apiClient.get(`/tasks/project/${projectId}?page=${page}&size=${size}&sort=${sort}`);
    return response.data;
  },

  getTasksByCategory: async (categoryId: number, page = 0, size = 10, sort = 'createdAt,desc'): Promise<PageResponse<TaskDto>> => {
    const response = await apiClient.get(`/tasks/category/${categoryId}?page=${page}&size=${size}&sort=${sort}`);
    return response.data;
  },

  getTasksByStatus: async (status: TaskStatus, page = 0, size = 10): Promise<PageResponse<TaskDto>> => {
    const response = await apiClient.get(`/tasks/status/${status}?page=${page}&size=${size}`);
    return response.data;
  },

  getTasksByPriority: async (priority: Priority, page = 0, size = 10): Promise<PageResponse<TaskDto>> => {
    const response = await apiClient.get(`/tasks/priority/${priority}?page=${page}&size=${size}`);
    return response.data;
  },

  getOverdueTasks: async (page = 0, size = 10): Promise<PageResponse<TaskDto>> => {
    const response = await apiClient.get(`/tasks/overdue?page=${page}&size=${size}`);
    return response.data;
  }
};

export default taskService; 